linkchecker.go

```diff
@@ -0,0 +1,437 @@
+	HTTPClient  *http.Client
+	Wg          sync.WaitGroup
+	Results     chan Result
+	output      io.Writer
+	errorLog    io.Writer
+	Scheme      string
+	Domain      string
+	Ratelimiter *rate.Limiter
+	CheckLink   CheckLink
+}
+
+type Option func(*LinkChecker) error
+
+func WithOutput(output io.Writer) Option {
+	return func(l *LinkChecker) error {
+		l.output = output
+		return nil
+	}
+}
+
+func WithErrorLog(errorLog io.Writer) Option {
+	return func(l *LinkChecker) error {
+		l.errorLog = errorLog
+		return nil
+	}
+}
+
+func WithConfigureRatelimiter(ratePerSec rate.Limit, burst int) Option {
+	return func(l *LinkChecker) error {
+		l.Ratelimiter = rate.NewLimiter(ratePerSec, burst)
+		return nil
+	}
+}
+
+func WithBufferedChannelSize(size int) Option {
+	return func(l *LinkChecker) error {
+		l.Results = make(chan Result, size)
+		return nil
+	}
+}
+
+func NewLinkChecker(opts ...Option) (*LinkChecker, error) {
+
+	linkchecker := &LinkChecker{
+		HTTPClient:  &http.Client{Timeout: 10 * time.Second},
+		output:      os.Stdout,
+		errorLog:    os.Stderr,
+		Ratelimiter: rate.NewLimiter(12, 24),
+		Results:     make(chan Result, 10),
+		CheckLink: CheckLink{
+			List: make(map[string]bool),
+		},
+	}
+
+	for _, o := range opts {
+		o(linkchecker)
+	}
+
+	return linkchecker, nil
+}
+
+func CheckSiteLinks(site string, opts ...Option) <-chan Result {
+	l, err := NewLinkChecker()
+	for _, o := range opts {
+		o(l)
+	}
+
+	if err != nil {
+		fmt.Fprintf(l.errorLog, "unable to create linkchecker struct, %s", err)
+	}
+
+	err = l.Check(site)
+	if err != nil {
+		fmt.Fprint(l.errorLog, err)
+	}
+
+	return l.Results
+
+}
+
+func (l *LinkChecker) Check(site string) error {
+
+	url, err := url.Parse(site)
+	if err != nil {
+		return err
+	}
+
+	l.Scheme, l.Domain = url.Scheme, url.Host
+
+	canonicalSite, err := l.CanonicaliseUrl(site)
+	if err != nil {
+		return err
+	}
+
+	referringSite := canonicalSite
+
+	l.Crawl(canonicalSite, referringSite)
+
+	close(l.Results)
+
+	return nil
+
+}
+
+func (l *LinkChecker) Crawl(site string, referringSite string) {
+
+	// get results for inital page
+	l.Wg.Add(1)
+	l.GetResult(site, referringSite)
+	l.Wg.Wait()
+
+	resp, err := l.GetResponse(site)
+	if err != nil {
+		fmt.Fprint(l.errorLog, err)
+	}
+
+	u, err := url.Parse(site)
+	if err != nil {
+		fmt.Fprint(l.errorLog, err)
+	}
+
+	// crawl pages and check links within domain
+	// and only those not in the CheckLink.List map
+	if u.Host == l.Domain && !l.IsCrawled(site) {
+
+		// add site to CheckLink.List map to handle
+		// link back (e.g pageA links pageB which links to pageA)
+		// add site in if statement to allow the top level page to
+		// be crawled.
+		l.AddSite(site)
+
+		// generate of list of links on page
+		links, err := l.ParseBody(resp.Body)
+		if err != nil {
+			fmt.Fprintf(l.errorLog, "unable to generate site list, %s", err)
+		}
+
+		for _, link := range links {
+
+			if !l.IsCrawled(link) {
+				l.Wg.Add(1)
+				go l.GetResult(link, site)
```

Comment:
The highest-level comment when I reviewed this the other day with Gabe and Kevin was that "Crawl should call itself". In other words, this task is _in principle_ recursive, isn't it? The most natural way to model that in Go would be with a function that calls itself recursively. Something like this, in pseudocode:

```
CRAWL url:
	if already checked:
		skip
	check HEAD status
	if non-local:
		return result
	else:
		fetch page
		for each link on page:
			CRAWL link
```